business/authentication/session_persistor.go

```diff
@@ -0,0 +1,335 @@
+		sessionDataBuffer.WriteString(base64SessionData)
+
+		for i := 1; i < numChunks; i++ {
+			cookieName := fmt.Sprintf("%s-aes-%d", config.TokenCookieName, i)
+			authChunkCookie, chunkErr := r.Cookie(cookieName)
+			if chunkErr != nil {
+				return nil, fmt.Errorf("failed to read session cookie chunk number %d: %w", i, chunkErr)
+			}
+
+			sessionDataBuffer.WriteString(authChunkCookie.Value)
+		}
+
+		// Get the concatenated session data
+		base64SessionData = sessionDataBuffer.String()
+	} else if chunksCookieErr != http.ErrNoCookie {
+		// Tolerate a "no cookie" error, but if error is something else, throw up the error.
+		return nil, fmt.Errorf("failed to read the chunks cookie: %w", chunksCookieErr)
+	}
+
+	// Persisted data has been read, but it's base64 encoded and it's also encrypted (per
+	// the process in CreateSession function). Reverse the encoding and, then, decrypt the data.
+	cipherSessionData, err := base64.StdEncoding.DecodeString(base64SessionData)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode session data: %w", err)
+	}
+
+	block, err := aes.NewCipher([]byte(config.GetSigningKey()))
+	if err != nil {
+		return nil, fmt.Errorf("error when restoring the session - failed to create the cipher: %w", err)
+	}
+
+	aesGCM, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, fmt.Errorf("error when restoring the session - failed to create gcm: %w", err)
+	}
+
+	nonceSize := aesGCM.NonceSize()
+	nonce, cipherSessionData := cipherSessionData[:nonceSize], cipherSessionData[nonceSize:]
+
+	sessionDataJson, err := aesGCM.Open(nil, nonce, cipherSessionData, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error when restoring the session - failed to decrypt: %w", err)
+	}
+
+	// sessionDataJson is holding the decrypted data as a string. This should be a JSON document. Let's parse it.
+	var sData sessionData
+	err = json.Unmarshal(sessionDataJson, &sData)
+	if err != nil {
+		return nil, fmt.Errorf("error when restoring the session - failed to parse the session data: %w", err)
+	}
+
+	// Check that the currently configured strategy matches the strategy set in the session.
+	// This is to prevent taking a session as valid if somebody re-configured Kiali with a different auth strategy.
+	if sData.Strategy != config.Get().Auth.Strategy {
+		log.Tracef("Session is invalid because it was created with authentication strategy %s, but current authentication strategy is %s", sData.Strategy, config.Get().Auth.Strategy)
+		p.TerminateSession(r, w) // Kill the spurious session
+
+		return nil, nil
+	}
+
+	// Check that the session has not expired.
+	// This is just a sanity check, because browser cookies are set to expire at this date and the browser
+	// shouldn't send expired cookies.
+	if !util.Clock.Now().Before(sData.ExpiresOn) {
+		log.Tracef("Session is invalid because it expired on %s", sData.ExpiresOn.Format(time.RFC822))
+		p.TerminateSession(r, w) // Clean the expired session
+
+		return nil, nil
```

Comment:
Should these return errors? I wonder if the client will still read the `SetCookie` header if the resp is not a 2xx.

Reply:
AFAIK, the `SetCookie` should be read even if the response is not 2xx.

But as the comment says, it should be just for sanity. The expiration time field of the sessionData and the browser cookie expiration time should both be synced. If for some reason the `SetCookie` is not read, the browser should still discard the cookies by it's own.

Only a forged cookie could have unsynced times. The benign case is a machine with a clock behind the real time, but still the caller should transparently see no active session rather than an error.

So IMHO, no error should be generated and the caller should simply "see" that there is no active session.